lib.nix

```diff
@@ -2,8 +2,39 @@
 # nix-repl lib.nix
 
 let
-  hostPkgs = import <nixpkgs> {};
-  lib = hostPkgs.lib;
+  # fetch nixpkgs and give the expected hash
+  unpackTar = input: let
+    nixcfg = import <nix/config.nix>;
+    script = ''
+      set -x
+      export PATH=$coreutils
+      $coreutils/mkdir $out
+      cd $out
+      $gzip -d < $input | $tar -x --strip-components=1
+    '';
+  in builtins.derivation {
+    system = builtins.currentSystem;
+    name = "${input.name}-unpacked";
+    builder = nixcfg.shell;
+    args = [ (builtins.toFile "builder" script) ];
+    inherit input;
+    coreutils = builtins.storePath nixcfg.coreutils;
+    tar = builtins.storePath nixcfg.tar;
+    gzip = builtins.storePath nixcfg.gzip;
+  };
+  fetchNixpkgsTarWithNix = let
+    spec = builtins.fromJSON (builtins.readFile ./nixpkgs-src.json);
+  in import <nix/fetchurl.nix> {
+    url = "https://github.com/${spec.owner}/${spec.repo}/archive/${spec.rev}.tar.gz";
```

Comment:
https://github.com/NixOS/nix/blob/master/src/libstore/builtins.cc#L56-L61
nar, optionally wrapped in xz, and no other formats allowed